### PR TITLE
PYIC-3894 Initial code changes for TiCF CRI stub to return TiCF VC. Added CF config files for dev and build env. And git workflow pipeline.

### PR DESCRIPTION
**Pull request by rpayal**
## Proposed changes
Creating a Stub of the TICF CRI

### What changed
- Created TICF stub
- Added ticf stub pipeline changes
- Added CF config for infrastructure creation and deployment of ticf stub

### Issue tracking

- [PYI-3894](https://govukverify.atlassian.net/browse/PYI-3894)

### Environment variables or secrets

- [ ] No environment variables or secrets were added or changed

- [ ] Documented in the [README](./blob/main/README.md)
- [ ] Added to deployment repository
- [ ] Added to local startup repository

### Other considerations

- [ ] Update [README](./blob/main/README.md) with any new instructions or tasks

### CODE REVIEW

**Comment by rpayal** [comment]
@vam-gds pls. review the pipeline config for the new stub along with CF deployment config.

File: di-ipv-ticf-stub/README.md
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/README.md]: I've not seen it stylised like this before - it's always just been 'TICF' to me.

File: di-ipv-ticf-stub/README.md
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/README.md:4]: It might be worth linking to the confluence page here as well, for background.

File: di-ipv-ticf-stub/core-dev-deploy/template.yaml
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/core-dev-deploy/template.yaml:125]: Should there be one of these per environment?

File: di-ipv-ticf-stub/lambdas/src/common/ssmParameter.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/ssmParameter.ts]: This means we'll just return `undefined` if fetching the parameter fails, which seems a bit surprising - is a missing parameter ever _not_ an error?

Perhaps we should be checking for undefined here instead, e.g.:
```
export const getSsmParameter = async (name: string): Promise<string> => {
  console.info(`Getting value for ssm parameter: ${name}`);
  // Retrieve a single parameter
  const parameter = await getParameter(name);
  if (parameter === undefined) {
    throw new Error(`Could not retrieve parameter: ${name}`);
  }
  return parameter;
};
```

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: I don't think you need the extra brackets here - could just be `body = { errorMessage: "Bad request." }`?

You could also inline this:
```
  _400: (body?: { [key: string]: any }) => ({
    statusCode: 400,
    body: JSON.stringify(body ?? { errorMessage: "Bad request" }, undefined, 2),
  }),
```

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: This could probably just be `object`?

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: If you've specified the type of the function (i.e. `Handler`), you don't need to specify the type of the event explicitly.

In our case, we can actually use a more specific handler type, to give us type-safety over the requests and responses: `APIGatewayProxyHandlerV2`.

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: In general we should use `const` for all variables, unless we need them to be mutable (but should try to avoid that)

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: If you want to apply a default, you can do it in one go. However, I would question whether we actually want a default here - do we ever expect the SSM parameters not to be set?

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: We shouldn't need the types here - TS should infer it for you :)

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: Similarly, why do we need to declare this separately to assignment? (ditto below)

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: We could probably pull out the evidence part, since the wrapper JWT is always the same

File: di-ipv-ticf-stub/lambdas/tsconfig.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/tsconfig.json:1]: Did you copy this from somewhere? I think we can probably simplify this a little - e.g. just including `src`, rather than excluding a bunch of stuff.

File: di-ipv-ticf-stub/openAPI/ticf-external.yaml
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/openAPI/ticf-external.yaml]: I'd be keen to give this a path if possible, just so it's easier to disambiguate from any other endpoints in future - perhaps `/risk-assessment`? (I don't know what the real one will look like yet)

File: di-ipv-ticf-stub/openAPI/ticf-external.yaml
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/openAPI/ticf-external.yaml]: Do we need to map the request, can we just pass the body through as-is?

File: di-ipv-ticf-stub/lambdas/.eslintrc.js
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/.eslintrc.js:1]: Does this work?

I think you'll need to use the `@typescript-eslint` plugin to handle typescript properly. You could look at the `delete-user-data` lambda in core-back as an example.

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/package.json]: Can we add a script to run eslint?

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/package.json]: Are we using `nodemon` or `ts-node`?

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/package.json]: Has this repo been moved to `govuk-one-login`?

File: di-ipv-ticf-stub/core-dev-deploy/template.yaml
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/core-dev-deploy/template.yaml:1]: I think it'd be good to get a review from one of the SREs on this (cc @vam-gds ), as there's a lot of stuff here and I'm not super close to our standards for it.

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: These two conditions are really mutually-exclusive - if the TICF API times out, then it cannot include any CIs.

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: As a general rule for lambda handlers, I'd like to separate logic to parse the request and serialize the response from business logic.

In this case, I think the handler should do validation for the incoming request and construction of the `ApiGatewayProxyResult`, but all the logic around generating the signed JWTs should live in a service module.

**Review by Joe-Edwards-GDS — COMMENTED** [review]
I think this is a good start, but I'm keen to set some solid patterns for TS lambdas, so I've left quite a few comments!

I think it's fine to use the SSM parameter approach as a way to test the core stub logic, but we need to make it dynamically configurable through a management API too.

Do you think we need to raise a separate sub-task for that change? Or we could even split out into a separate task to try and unblock the main TICF lambda development.

File: di-ipv-ticf-stub/README.md
Comment (rpayal) [reply di-ipv-ticf-stub/README.md]: done

File: di-ipv-ticf-stub/README.md
Comment (rpayal) [reply di-ipv-ticf-stub/README.md:4]: done

File: di-ipv-ticf-stub/core-dev-deploy/template.yaml
Comment (rpayal) [reply di-ipv-ticf-stub/core-dev-deploy/template.yaml:125]: done

File: di-ipv-ticf-stub/lambdas/src/common/ssmParameter.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/common/ssmParameter.ts]: done

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: done

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: removed nodemon.....ts-node is needed to access environment variable.

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: will check.

File: di-ipv-ticf-stub/core-dev-deploy/template.yaml
Comment (rpayal) [reply di-ipv-ticf-stub/core-dev-deploy/template.yaml:1]: yeah already added @vam-gds for review of pipeline and CF deployment config

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: done

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: done

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: moved....so now using govuk-one-login.

**Comment by rpayal** [comment]
> I think this is a good start, but I'm keen to set some solid patterns for TS lambdas, so I've left quite a few comments!
> 
> I think it's fine to use the SSM parameter approach as a way to test the core stub logic, but we need to make it dynamically configurable through a management API too.
> 
> Do you think we need to raise a separate sub-task for that change? Or we could even split out into a separate task to try and unblock the main TICF lambda development.

Yeah will be good to handle management api approach through a separate ticket. Do we need by storing value for timeout etc as a db value ?

**Comment by Joe-Edwards-GDS** [comment]
> Yeah will be good to handle management api approach through a separate ticket. Do we need by storing value for timeout etc as a db value ?

I don't think we need to worry about the timeout exactly, but I'd expect DynamoDB to have records _something_ like:
| userId | evidence | ttl |
| --- | --- | --- |
| urn:uuid:some-user-id | `{ "type": "RiskAssessment" }` | 1700667070 |
| urn:uuid:some-other-id | `{ "type": "RiskAssessment", "ci": ["XXX"], "txt": "some-txn-id" }` | 1700667070 |

I'll raise a separate ticket for this

File: di-ipv-ticf-stub/lambdas/.eslintrc.js
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/.eslintrc.js:1]: I think this still needs updating to use the typescript plugin - have a look at the example in the delete-user-data lambda.

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [reply di-ipv-ticf-stub/lambdas/package.json]: Can you elaborate on this? I can't see any usages of it in the code

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: I think you could just use `body: object` here?

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: You should always use `===` in typescript, unless you _really_ want type coercion.

In this case I don't think these would ever be `null`, but it's perfectly idiomatic to use falsiness - assuming that 'empty' is invalid:
```
if (!ticfRequest || !ticfRequest.vot) {
    return apiResponses._400({ errorMessage: "Invalid request" });
  }
```
Should we also check the other properties of the request are present too?

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: It would be nice to specify the return types here - I think it's `ApiGatewayProxyResultV2`?

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts:7]: Do you need to include a content type here - i.e. `headers: { 'content-type': 'application/json' },`?

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: I don't think this needs to be a class - it doesn't have any state. In general I prefer quite functional Typescript anyway*, and I think this could just be a direct export:

```
export async function processGetVcRequest(ticfRequest: TicfRequest): Promise<TicfResponse> {
  ...
}
```
*JavaScript classes are _weird_ and full of gotchas

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: This is swallowing the underlying error - can we just let that bubble up? An error message describing the parameter being fetched feels more useful anyway.

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: You shouldn't need the explicit type annotation on this, it should be inferred

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: I don't think this can ever happen, now that `getSsmParameter` throws for an undefined parameter

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: Can just be
```
const buildJwtParams = {
  ...
};
```

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: Can just be
```
const returnJwt = await buildSignedJwt(buildJwtParams);

return {
  sub: ticfRequest.sub,
  ...
};
```

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: Do you think this needs to be configurable as well? How does it work for the other stubs?

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/services/ticfService.ts:108]: There's still repetition here, that I think we can avoid?
```
function getCustomClaims(
  timeOutVC: boolean,
  includeCIToVC: boolean,
  userId: string
): JWTPayload {
  return {
    sub: userId,
    iss: process.env.ISSUER,
    aud: "https://development-di-ipv-core-front.london.cloudapps.digital",
    nbf: Date.now(),
    vc: {
      evidence: [getEvidenceItem(timeOutVC, includeCIToVC)],
      type: ["VerifiableCredential", "RiskAssessmentCredential"],
    },
    jti: "urn:uuid:" + uuid(),
  };
}

function getEvidenceItem(
  timeOutVC: boolean,
  includeCIToVC: boolean
): TicfEvidenceItem {
  if (timeOutVC) {
    return { type: "RiskAssessment" };
  } else {
    if (includeCIToVC) {
      return {
        type: "RiskAssessment",
        ci: ["V03"],
        txn: uuid(),
      };
    } else {
      return {
        type: "RiskAssessment",
        txn: uuid(),
      };
    }
  }
}
```
This should also make it easy to update when the evidence block comes from the management API

File: di-ipv-ticf-stub/lambdas/src/domain/ticfResponse.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/src/domain/ticfResponse.ts:1]: (If these types start proliferating, then I think it'd also be fine to put them in a single file)

File: di-ipv-ticf-stub/lambdas/.eslintrc.js
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/.eslintrc.js]: Can we try to keep this if possible - where do you need to use `any`?

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: in service using - process.env.TICF_PARAM_BASE_PATH etc.   when removing ts-node it was complaining.

File: di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/handlers/ticfHandler.ts]: check added for other properties also.

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: already taken care

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/services/ticfService.ts:108]: done

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/package.json]: Are you planning to write integration tests for this?

File: di-ipv-ticf-stub/lambdas/package.json
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/package.json]: It looks like axios provides its own types now, so you might not need this

File: di-ipv-ticf-stub/lambdas/test/unit/ticfHandler.test.ts
Comment (Joe-Edwards-GDS) [inline di-ipv-ticf-stub/lambdas/test/unit/ticfHandler.test.ts:1]: If you're having type errors in your test, you might need to tell tsconfig to include them, which either means:
- Removing the exclusion on `node_modules` (so it includes the Jest types by default)
- Including them explicitly, with `"types": "jest"`

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: will remove - earlier added was trying to create integration test agnst local run.

File: di-ipv-ticf-stub/openAPI/ticf-external.yaml
Comment (rpayal) [reply di-ipv-ticf-stub/openAPI/ticf-external.yaml]: https://docs.aws.amazon.com/apigateway/latest/developerguide/integration-passthrough-behaviors.html
- The request payload is transformed using the template.
Changes application/x-www-form-urlencoded to application/json

Comment (Joe-Edwards-GDS) [reply di-ipv-ticf-stub/openAPI/ticf-external.yaml]: Do we need this at all though? I think we can just pass through directly to the lambda

File: di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/common/apiResponses.ts]: Done

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: done

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: done

**Review by Joe-Edwards-GDS — DISMISSED** [review]
Happy to get this unblocked now :)

I will probably raise another PR to tidy up a few extra bits and pieces (e.g. sorting out this `di-stub-oauth-client` issue), but no need to delay any further.

File: di-ipv-ticf-stub/lambdas/package.json
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/package.json]: used for integ test - removed now.

File: di-ipv-ticf-stub/lambdas/.eslintrc.js
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/.eslintrc.js]: changed to 'error'

**Comment by rpayal** [comment]
Tested in dev.

**Review by Joe-Edwards-GDS — DISMISSED** [review]

File: di-ipv-ticf-stub/lambdas/src/services/ticfService.ts
Comment (rpayal) [reply di-ipv-ticf-stub/lambdas/src/services/ticfService.ts]: Updated to get componentId from ssm for VC's aud value.

**Review by Joe-Edwards-GDS — APPROVED** [review]

**Pull request merged**
